src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php

```diff
@@ -0,0 +1,133 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Config\Builder;
+
+use Symfony\Component\Config\Definition\ArrayNode;
+use Symfony\Component\Config\Definition\BaseNode;
+use Symfony\Component\Config\Definition\BooleanNode;
+use Symfony\Component\Config\Definition\EnumNode;
+use Symfony\Component\Config\Definition\FloatNode;
+use Symfony\Component\Config\Definition\IntegerNode;
+use Symfony\Component\Config\Definition\NodeInterface;
+use Symfony\Component\Config\Definition\NumericNode;
+use Symfony\Component\Config\Definition\PrototypedArrayNode;
+use Symfony\Component\Config\Definition\ScalarNode;
+use Symfony\Component\Config\Definition\StringNode;
+use Symfony\Component\Config\Definition\VariableNode;
+
+/**
+ * @author Alexandre Daubois <alex.daubois@gmail.com>
+ *
+ * @internal
+ */
+final class ArrayShapeGenerator
+{
+    public static function generate(NodeInterface $node): string
+    {
+        return str_replace("\n", "\n * ", self::doGeneratePhpDoc($node));
+    }
+
+    private static function doGeneratePhpDoc(NodeInterface $node, int $nestingLevel = 1): string
+    {
+        if (!$node instanceof ArrayNode) {
+            return match (true) {
+                $node instanceof BooleanNode => 'bool',
+                $node instanceof StringNode => 'string',
+                $node instanceof NumericNode => self::handleNumericNode($node),
+                $node instanceof EnumNode => $node->getPermissibleValues('|'),
+                $node instanceof ScalarNode => 'string|int|float|bool|null',
+                $node instanceof VariableNode => 'mixed',
+            };
+        }
+
+        if ($node instanceof PrototypedArrayNode) {
+            $isHashmap = (bool) $node->getKeyAttribute();
+
+            return 'array<'.($isHashmap ? 'string, ' : '').self::doGeneratePhpDoc($node->getPrototype(), 1 + $nestingLevel).'>';
```

Comment:
When the node is not an hashmap, I would suggest documenting it as `list<T>` instead of `array<T>`.

`array<T>` accepts _any_ keys, but the array keys will be lost as soon as merging needs to happen as the merging will be performed as a list merging. The fact that keys are preserved without merging is a side effect of a micro-optimization done early in the component implementation (and which I'd like to deprecate as it makes it easy to hide configuration mistakes when the simple usage of a misconfigured map will work only because no merging is performed).